File: S1/Bicicleta.java
package S1;

public abstract class Bicicleta extends Thread {

	private final long dorsal;
	static final int TIEMPO_CARRERA = 60;
	private int retirada;
	
	Bicicleta()
	{
		super();
		dorsal = this.getId();
		retirada = TIEMPO_CARRERA;
	}

	void setRetirada(int valor){
		this.retirada = valor;
	}

	long getDorsal()
	{
		return dorsal;
	}

	@Override
	public String toString() {
		return "Bicicleta de " + getTipo();
	}

	abstract String getTipo();

	@Override
	public void run()
	{
		for(int i = 1; i <= TIEMPO_CARRERA; i++)
		{
			if(i == TIEMPO_CARRERA)
				System.out.println("Bici nº " + dorsal + " ha terminado la carrera de " + getTipo());
			else if(i != retirada)
				System.out.println("Bici nº " + dorsal + " está corriendo en la carrera de " + getTipo());
			else
			{
				System.out.println("La bici " + dorsal + " ha sido descalificada de la carrera de " + getTipo());
				//habria que poner aqui stop ya que no lo hacemos en Carrera??
				this.stop();
				//aunque no la estariamos quitando del Array, es la unica pega que le veo a esto, pero realmente hace falta quitarlo del array?
			}

			try {
				Thread.sleep(1000);
			} catch (InterruptedException e){
				System.out.println("No se ha podido completar la espera");
			}
		}
	}
}

File: P1/S1-voluntario/Bicicleta.rb
module S1
	public
	class Bicicleta

		@@TIEMPO_CARRERA = 60

		def initialize(id)
			@retirada = @@TIEMPO_CARRERA
			@dorsal = id
		end

		def setRetirada(valor)
			@retirada = valor
		end

		def self.TIEMPO_CARRERA
			@@TIEMPO_CARRERA
		end

		def getDorsal
			@dorsal
		end

		attr_reader :retirada

		def to_s
			"Bicicleta de " + getTipo
		end
		
		def getTipo
			#metodo abstracto
		end

		def setId(id)
			@dorsal = id
		end

		def run()
			
			for i in 1..@@TIEMPO_CARRERA do

				if i == @@TIEMPO_CARRERA
					puts "Bici nº " + @dorsal.to_s + " ha terminado la carrera de " + getTipo()
					#cotinuar = false
					break
				elsif i != @retirada
					puts "Bici nº " + @dorsal.to_s + " está corriendo en la carrera de " + getTipo()
				else
					puts "La bici " + @dorsal.to_s + " ha sido descalificada de la carrera de " + getTipo()
					#cotinuar = false
					break
				end
				sleep(1)
			end

		end

	end
end
